server/src/main/java/org/tctalent/server/service/db/impl/SalesforceServiceImpl.java

```diff
@@ -411,6 +411,25 @@ public void createOrUpdateCandidateOpportunities(
                 closingCommentsForCandidate = candidateOppParams.getClosingCommentsForCandidate();
                 employerFeedback = candidateOppParams.getEmployerFeedback();
                 relocationInfo = candidateOppParams.getRelocationInfo();
+
+                // If relocationInfo not already included in params and new stage is 'Relocated',
+                // update the SF case relocation info - just to assist with monitoring & evaluation,
+                // a failsafe in case admin users haven't clicked the 'Update case stats' button
+                // when updating relocating dependant info, which can be set on a visa job check
+                // or directly on the Candidate Opp via the 'Upload' tab.
+                // Typically, would use CandidateOpportunityService here, but that would create a
+                // dependency cycle between beans — so instead querying the SalesforceJobOpp to get
+                // the CandidateOpportunity required for processSfCaseRelocationInfo()
+                if (relocationInfo == null && stage == CandidateOpportunityStage.relocated) {
+                    Optional<CandidateOpportunity> candidateOpp =
+                        jobOpportunity.getCandidateOpportunities()
+                            .stream()
+                            .filter(opp -> opp.getCandidate().getId().equals(candidate.getId()))
+                            .findFirst();
+                    if (candidateOpp.isPresent()) {
+                        relocationInfo = processSfCaseRelocationInfo(candidateOpp.get(), candidate);
+                    }
+                }
```

Comment:
It seems like there should be an easier way to do this. Wouldn't it be easier to retrieve the required TC candidate from the TC JobOpp - matching candidate.sfId with the opp.getCandidate().getId().

Then just fetch the SF candidateOpp using findCandidateOpportunity.

Reply:
@camerojo the problem is that processSfCaseRelocationInfo requires a Candidate Opportunity, whereas findCandidateOpportunity returns an Opportunity. I might be able to change the typing on processRelocationInfo, I'll take a look, but otherwise this seemed to be my only option.

Reply:
Ahhh - gotcha